x-pack/plugins/security_solution/public/common/containers/source/index.tsx

```diff
@@ -94,87 +97,72 @@ export const WithSource = React.memo<WithSourceProps>(({ children, indexToAdd, s
     return configIndex;
   }, [configIndex, indexToAdd]);
 
-  return (
-    <Query<SourceQuery.Query, SourceQuery.Variables>
-      query={sourceQuery}
-      fetchPolicy="cache-first"
-      notifyOnNetworkStatusChange
-      variables={{
-        sourceId,
-        defaultIndex,
-      }}
-    >
-      {({ data }) =>
-        children({
-          indicesExist: get('source.status.indicesExist', data),
-          browserFields: getBrowserFields(
-            defaultIndex.join(),
-            get('source.status.indexFields', data)
-          ),
-          indexPattern: getIndexFields(defaultIndex.join(), get('source.status.indexFields', data)),
-        })
-      }
-    </Query>
-  );
-});
-
-WithSource.displayName = 'WithSource';
-
-export const indicesExistOrDataTemporarilyUnavailable = (indicesExist: boolean | undefined) =>
-  indicesExist || isUndefined(indicesExist);
-
-export const useWithSource = (sourceId: string, indices: string[]) => {
   const [loading, updateLoading] = useState(false);
-  const [indicesExist, setIndicesExist] = useState<boolean | undefined | null>(undefined);
-  const [browserFields, setBrowserFields] = useState<BrowserFields | null>(null);
-  const [indexPattern, setIndexPattern] = useState<IIndexPattern | null>(null);
-  const [errorMessage, updateErrorMessage] = useState<string | null>(null);
+  const [{ browserFields, errorMessage, indicesExist, indexPattern }, setState] = useState<
+    UseWithSourceState
+  >({
+    indicesExist: undefined,
+    browserFields: EMPTY_BROWSER_FIELDS,
+    errorMessage: null,
+    indexPattern: getIndexFields(defaultIndex.join(), []),
+  });
 
   const apolloClient = useApolloClient();
-  async function fetchSource(signal: AbortSignal) {
-    updateLoading(true);
-    if (apolloClient) {
-      apolloClient
-        .query<SourceQuery.Query, SourceQuery.Variables>({
-          query: sourceQuery,
-          fetchPolicy: 'cache-first',
-          variables: {
-            sourceId,
-            defaultIndex: indices,
-          },
-          context: {
-            fetchOptions: {
-              signal,
-            },
-          },
-        })
-        .then(
-          (result) => {
-            updateLoading(false);
-            updateErrorMessage(null);
-            setIndicesExist(get('data.source.status.indicesExist', result));
-            setBrowserFields(
-              getBrowserFields(indices.join(), get('data.source.status.indexFields', result))
-            );
-            setIndexPattern(
-              getIndexFields(indices.join(), get('data.source.status.indexFields', result))
-            );
-          },
-          (error) => {
-            updateLoading(false);
-            updateErrorMessage(error.message);
-          }
-        );
-    }
-  }
 
   useEffect(() => {
     const abortCtrl = new AbortController();
```

Comment:
I think we are missing something fundamental here, to avoid to update the state if the promise has been canceled.

https://github.com/facebook/react/issues/14326#issuecomment-472043812 or like here in our code https://github.com/elastic/kibana/blob/master/x-pack/plugins/security_solution/public/alerts/containers/detection_engine/alerts/use_privilege_user.tsx#L40